_posts/2023-03-13-llvm-opaque-pointers.md

```diff
@@ -0,0 +1,85 @@
+
+```llvm
+; Function Attrs: uwtable(sync)
+define internal i32 @"*Foo#initialize<Int32>:Int32"(ptr %self, i32 %x) #0 {
+entry:
+  %0 = getelementptr inbounds %Foo, ptr %self, i32 0, i32 1
+  store i32 %x, ptr %0, align 4
+  ret i32 %x
+}
+```
+
+To get to there, a few LLVM instructions have to be built differently depending on whether typed or opaque pointers are used, and the `getelementptr` instruction is one such example. The object type used to be inferred from the given pointer, but now that opaque pointers do not carry that information, the IR generator — our Crystal compiler — needs to supply this object type separately. In this case, the `@x` instance variable and the `#initialize` method both belong to `Foo`, so Crystal knows to pass `Foo` to `getelementptr`. But this instruction is also used in a few dozen other places in the compiler, for which no universal migration is applicable.
+
+Work on the migration to opaque pointers started in October 2022, a month after LLVM 15's release, and after countless segfaults and spec failures, [Crystal now supports LLVM 15 on the master branch](https://github.com/crystal-lang/crystal/pull/13173). As it was a rather huge effort, surely we want the opaque pointers to deliver the performance benefits that LLVM promises. So here are some numbers collected from re-building the compiler itself on an Apple M2, first with an LLVM 14 compiler, then with an LLVM 15 one:
+
+* Non-release build:
+  * Codegen (crystal): 2.65s → 2.84s
+  * Codegen (bc+obj): 5.91s → 5.20s
+  * Codegen (linking): 0.76s → 0.34s
+  * dsymutil: 0.35s → 0.39s
+* Release build:
+  * Codegen (bc+obj): 247.86s → 184.37s
+  * Codegen (linking): 0.45s → 0.33s
+  * dsymutil: 0.63s → 0.52s
+
+If we consider only the last 3 stages, which are fully in LLVM's control, that's a 18% speed-up for non-release builds and 34% speed-up for release builds! Similar figures were reported by developers who were eager enough to re-build Crystal with LLVM 15. Although it takes 0.2 second more on average to generate the LLVM IR for a program as large as Crystal itself, LLVM's improvement outweighs it by a large margin. That migration effort has certainly paid off.
+
+## How does this affect me?
+
+If your Crystal projects do not deal with LLVM functionality directly, there is no action to take; the official Crystal distributions will soon start using LLVM 15 to build Crystal. Those compilers will use opaque pointers, whereas ones built with LLVM 8 - 14 will continue to use typed pointers.
```

Comment:
I think it's worth noting again that all users are positively affected by increased build performance.
Maybe some repositioning can help make this a bit more clearer:

```suggestion
The official Crystal distributions will soon start using LLVM 15 to build Crystal. Those compilers will use opaque pointers and show improvements in codegen time. Compilers built with LLVM 14 and below will continue to use typed pointers.

If your Crystal project uses the stdlib's LLVM API directly, there are some deprecations to note. Otherwise this change does not affect Crystal programs in any way. It just speeds up the compiler.
```